ITE/6/README.md

```diff
@@ -0,0 +1,574 @@
+    will support the agreed-upon format once finalized. This way we have a
+    single ecosystem of software supply chain security.
+
+Functional requirements:
+
+*   Must support user-defined types and schemas, for two reasons:
+    *   To allow in-toto users to more naturally express attestations, as
+        explained above.
+    *   Because Binary Authorization does not want to require its users to use
+        the existing in-toto link schema, which is overly specific.
+*   Should allow indexing of attestations by artifact ID, without having to
+    understand the user-defined schema.
+    *   Primary reason: To support generic attestation indexing/storage/fetching
+        without requiring user configuration for every type of attestation.
+    *   Secondary reason: To simplify the programming model of policies. The
+        binding between artifact and attestation can be done in a framework
+        without requiring type-dependent configuration.
+    *   Implication: the association between attestation and primary artifact ID
+        must be standardized.
+*   Should allow identification of related artifacts IDs given an attestation,
+    without having to understand the user-defined schema.
+    *   Reason: To support "inline attestations," where the client fetches and
+        sends all required attestations to the server for verification. The
+        client does not know the policy ahead of time or understand all
+        attestation types.
+    *   Example: Given a provenance attestation for a docker image, it should be
+        possible to identify all the materials generically.
+    *   Implication: the association between attestation and related artifact
+        IDs must be standardized.
+
+Nonfunctional requirements:
+
+*   Must support backwards compatible links that can be consumed by existing
+    layout files.
+*   Must differentiate between different types of related artifacts (only if
+    related artifacts are standardized.) Examples: materials vs products,
+    sources vs build tools.
+    *   Should be type-dependent, rather than mandating "materials" and
+        "products."
+
+# Reasoning
+
+## Reason for `subject` and `materials` to be standard Attestation fields
+
+There are two main reasons for standardizing `subject` and `materials` within
+Attestation schema.
+
+First, doing so allows policy engines to make decisions without requiring
+`attestation_type`-specific logic or configuration. Binary Authorization
+policies today are purely about "does an attestation exist that is signed by X
+with subject Y", and similarly in-toto layouts are about "does an attestation
+exist that is signed by X with materials/products Z?"[1] These relatively simple
+policies are quite powerful. With this proposal, such policies become more
+expressive without any additional configuration: "does an attestation exist that
+is signed by X having type T, with subject Y and/or materials Z?"
```

Comment:
Another way of looking at this: an attestation is a partial definition of a directed graph, in which edges connect sources materials to artifacts (to use your terminology), and the edges themselves are labelled with a pointer to the attestation that introduced them. Then these attestation compose naturally, and it is also possible to traverse this graph forwards or backwards, also taking into account the trust model in the various principals that sign these attestations.

This also produces a nice way of formulating what a "root of trust" is: from a given node in the graph, follow links backwards for attestations that are trusted (e.g. "GitHub actions"), until you reach an attestation that you cannot decompose any further (because you don't trust the upstream attestations).

If we go with this semantics, effectively an attestation with a set of N materials and a set of M subjects corresponds to (N * M) edges that contribute to the graph. This assumes all the materials contribute to all the subjects, which I think is a safe assumption; if this is not desired, then perhaps the attestation should be denormalized into individual edges, in order to offer more precision, at the expense of higher verbosity and lower entropy for the more common cases.

Reply:
BTW I think this also would connect nicely with @dlorenc 's ideas on how to incorporate more expressive entries in sigstore / rekor that contain links to / from other log entries.

Reply:
> BTW I think this also would connect nicely with @dlorenc 's ideas on how to incorporate more expressive entries in sigstore / rekor that contain links to / from other log entries.

That's no coincidence! @SantiagoTorres was behind much of that thinking :)

Reply:
Please take a look at the latest design.

1. `materials` is no longer in the Statement layer. Instead, it is only in the Provenance-type predicate. This means that we can more easily change it in the future. The downside means that you can't traverse the graph without understanding the Predicate type. However, the value of being able to traverse the graph without actually understanding the edge meanings is dubious. I can believe that perhaps there is a use case out there, but until we have a concrete one, our plan is to push as much down into the Predicate layer as possible. This allows each Predicate type to do what is natural in that domain. Also, things like SPDX have their own language for links, and our model now supports that.

2. Within Provenance, `materials` is an array instead of a map, so (1) the reference is by index instead of name, so there is less duplication, and (2) we have richer objects so you can attach more properties to them.

I don't follow the connection to Rekor. The big thing is standardizing `subject` so that you know what artifact the attestation is about. The links would be processed by the consumer, and they'd do further queries to walk the graph.

Reply:
Thanks Mark!

I think there is some value in representing this as a graph that can be traversed without understanding the meaning of the edges; e.g. in Oak we may want a way to recursively download and store all the attestations starting from a given artifact, and spanning the graph (perhaps up to a certain depth). This would form a sort of collection of facts about the binary, as well as its dependencies, and so on, which may be stored offline. Then a policy engine may take this as input and decide whether or not to accept the given binary (e.g. whether to run it or not).

Note that the component that does the downloading / traversal may not know anything at all about the meaning of the predicates, and the policy engine may not be able to perform queries on line.

Additionally, IMO I think there is a nice symmetry in considering each material (or set of materials, if that helps) as a directed edge in a graph (from / to the subject), in which the predicate is effectively an attribute of that edge. This way the topology of the graph should be traversable (and even cacheable, as per my previous paragraph) even without having to know anything about the individual predicates. :)

Reply:
I can see the prospects of this idea, but it's not for v1. :-) We're looking to release v1 within the next two weeks, and there are just too many unknowns to make the links part of the Statement layer.

Our bar for putting something in the Statement layer is that there must be widespread, concrete use cases showing the need to have the field be common for all attestations. Right now, the use cases for links in the Statement layer are still theoretical.

Some challenges:

- How do you terminate the graph traversal? The graph in reality is intractably large. For example, every build likely uses 10+ artifacts, each of which uses 10+, and so on, to a really large depth. Just terminating by depth is likely not the right choice, and even then, it is a huge number of attestations. You really do need to understand the predicates to do that effectively. Alternatively, you need to figure out a model that doesn't require this deep traversal (which is what I'm currently thinking for SLSA.)

- What is the model? Right now we have Statement = Subject + Predicate. Though simple, it is easy to reason about and matches natural language. If we instead put all the graph edges in the Statement layer, we'd likely need to use a different analogy. I'm not sure that a property graph (which is what you describe) is really the right model - it doesn't easily express Provenance, for examlpe. Instead it is better represented as a hypergraph or a bipartite graph (artifacts + attestations). I **do** want to express the graph nature in the model, but it will take much more time than we have available for v1.

- Will this support all predicates easily? For example, in the current model we can simply use SPDX as a predicate and it works (though not particularly cleanly). With the links in the Statement layer, now SPDX would have to simply ignore those links (making the links feature useless) or we'd need to figure out how to rework SPDX to fit the new model (complicating onboarding).

To be clear, I share your gut feeling that it's the "right" design choice, but for v1 we'll leave it to the Predicate and see what type of conventions emerge. Once we have real-world data, we can reevaluate.

Reply:
For the record, I filed a dedicated issue for this: https://github.com/in-toto/attestation/issues/6. Let's move any further discussion to that issue.